### PR TITLE
Initial testing PR

### PR DESCRIPTION
**Pull request by jn9e9**
Add low level interface tests.  Add initial operation client data driven tests and test generator.

Change content type to zero to match rust code.

Partial completion of https://github.com/parallaxsecond/parsec-client-go/issues/12

Signed-off-by: John 9e9 <jn9e9e9@gmail.com>

### CODE REVIEW

File: interface/operations/test/operations_client_test.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+type TestCase struct {
+	Name                  string      `json:"name"`
+	RequestData           interface{} `json:"request_data"`
+	ExpectedRequestBinary string      `json:"expected_request_binary"`
+	ResponseBinary        string      `json:"response_binary"`
+	ExpectedResponseData  interface{} `json:"expected_response"`
+	ExpectSuccess         bool        `json:"expect_success"`
+}
```
Comment (hug-dev) [inline interface/operations/test/operations_client_test.go:59]: Just for my understanding, with those tests you check that the stack correctly performs the conversions:
* from high-level functions with specific parameters (in `RequestData`) to binary data transmitted
* from the binary response to the high-level functions return value

File: interface/requests/mod.go [hunk 2]
```diff
@@ -37,3 +35,7 @@ const (
 	OpPsaVerifyMessage     OpCode = 0x0019
 	OpListKeys             OpCode = 0x001A
```
Comment (hug-dev) [inline interface/requests/mod.go:36]: Maybe we need a warning notice here, or a test of some sort, to also modify the `IsValid` method is an opcode bigger than the biggest one is added.
Same for all the `IsValid` methods 😬

**Review by hug-dev — COMMENTED** [review]
Thanks a lot for adding all this 🤩 
Let's start a discussion for the review. On a high-level view, this PR adds:

1. a Rust tool crate to generate JSON test cases from `NativeOperation`, `NativeResult` and their serialisations
2. Go test files checking that for each operation, the correct serialisation/deserialisation happens correctly on both dorection
3. Some adjustments to the wire protocol
4. Tests on the wire protocol header (de)serialisation
5. Tests on the authenticators and the (de)serialisation of the ID

Is that correct? Feel free to add more independant categories if I missed things.

1 and 2 seem to be the big pieces which might need more discussion. The rest seems to be easier to review and get merged. Where I want to go is that if we come with things in this PR that are independant with 1 and 2, perhaps you could send a separate PR for those, which we first agree and merge. Then we can focus the discussion on the meaty stuff!

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -53,11 +53,11 @@ jobs:
...
-      - name: Upload Coverage report to CodeCov
-        uses: codecov/codecov-action@v1.0.0
-        with:
-          token: ${{secrets.CODECOV_TOKEN}}
-          file: ./coverage.txt
+      # - name: Upload Coverage report to CodeCov
+      #   uses: codecov/codecov-action@v1.0.0
+      #   with:
+      #     token: ${{secrets.CODECOV_TOKEN}}
+      #     file: ./coverage.txt
```
[the remark this answers is not in the document]
Comment (jn9e9) [reply .github/workflows/build.yaml:60]: yes, that'd be good if somebody could sign up with codecov.io account for the project, get a token then make it available.

I have it on my fork, but have disabled it for the PR as it causes the build to fail without the secret set.

File: .gitignore [hunk 4]
```diff
@@ -1,3 +1,4 @@
 node_modules
 .vscode
 ...
+test-notes.md
```
[the remark this answers is not in the document]
Comment (jn9e9) [reply .gitignore:4]: nope, good spot, that was a file i was keeping track of the tests with.... will remove

File: interface/operations/operations_client.go [hunk 5]
```diff
@@ -82,6 +82,17 @@ func InitClient() (*Client, error) {
 	return client, nil
 }
 
+// InitClient initializes a Parsec client
+func InitClientFromConnection(conn connection.Connection) (*Client, error) {
+	client := &Client{
+		conn:     conn,
+		provider: requests.ProviderCore,
+		authType: auth.AuthUnixPeerCredentials,
```
[the remark this answers is not in the document]
Comment (jn9e9) [reply interface/operations/operations_client.go:90]: yes they would

File: interface/operations/test/operations_client_test.go (hunk 1 above)
Comment (jn9e9) [reply interface/operations/test/operations_client_test.go:59]: yes, that's correct, but a comment would be a good idea ;)

File: .github/workflows/build.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jn9e9) [reply .github/workflows/build.yaml:60]: have created https://github.com/parallaxsecond/parsec-client-go/issues/27

**Comment by jn9e9** [comment]
> 
> 
> Thanks a lot for adding all this 🤩
> Let's start a discussion for the review. On a high-level view, this PR adds:
> 
>     1. a Rust tool crate to generate JSON test cases from `NativeOperation`, `NativeResult` and their serialisations
> 
>     2. Go test files checking that for each operation, the correct serialisation/deserialisation happens correctly on both dorection
> 
>     3. Some adjustments to the wire protocol
> 
>     4. Tests on the wire protocol header (de)serialisation
> 
>     5. Tests on the authenticators and the (de)serialisation of the ID
> 
> 
> Is that correct? Feel free to add more independant categories if I missed things.
Yes, that is about it
> 
> 1 and 2 seem to be the big pieces which might need more discussion. The rest seems to be easier to review and get merged. Where I want to go is that if we come with things in this PR that are independant with 1 and 2, perhaps you could send a separate PR for those, which we first agree and merge. Then we can focus the discussion on the meaty stuff!
OK, I'll see if i can create a new branch and pull the 3-5 changes across and resubmit another PR to cover those.

**Comment by jn9e9** [comment]
Did wonder if the test data generator (and the test data) would be best in the parsec-interface-rs repo or its own.  We could either do it now, or wait until I've done a lump more testing and see how it evolves first?

**Comment by hug-dev** [comment]
> OK, I'll see if i can create a new branch and pull the 3-5 changes across and resubmit another PR to cover those.

That sounds good!

> Did wonder if the test data generator (and the test data) would be best in the parsec-interface-rs repo or its own. We could either do it now, or wait until I've done a lump more testing and see how it evolves first?

That is an interesting question... Let's open that one on the community channel to see what everyone thinks.

**Comment by hug-dev** [comment]
> Let's open that one on the community channel to see what everyone thinks.

See the linked issue for a proposal 😸 If you can set aside a few hours to read it

**Comment by jn9e9** [comment]
going to close this as will be ote with acceptance/implementation of this: https://github.com/parallaxsecond/parsec/issues/350

Retaining branch on fork for reference

**Pull request closed without merging**